### ISSUE TITLE
Casting `List[str]` to `List[Union[str, int]]` is not allowed by pyright

### ISSUE THREAD

**Issue by dennis-loh** [microsoft/pyright#2782]
**Describe the bug**
Pyright reporting error when we try to casting `List[str]` to `List[Union[str, int]]`

**To Reproduce**
Steps to reproduce the behavior.

```py
# Allowed
list1: str = ''
list2: Union[str, int] = list1



# Not allowed
list1: List[str] = []
list2: List[Union[str, int]] = list1

'''
Expression of type "list[str]" cannot be assigned to declared type "List[str | int]"
  TypeVar "_T@list" is invariant
    Type "str" cannot be assigned to type "str | int"
      "str" is incompatible with "int"
'''
```

**Expected behavior**
Pyright should not report any error.

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool? Which version? You can find the version of the VS Code extension by clicking on the Pyright icon in the extensions panel.

VSCode: 1.64.0-insider
Pyright: v1.1.202

**Additional context**

**Comment by erictraut** [comment]
This is the correct behavior. The type parameter for `List` and other mutable container classes is "invariant" which means that you cannot substitute `str` for `Union[str, int]` or vice versa. This is a common source of confusion for people who are new to the Python type system. For details, please refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#generic-types).
